Realm/RLMSyncUser.mm

```diff
@@ -383,6 +368,65 @@ - (void)revokePermission:(RLMSyncPermission *)permission callback:(RLMPermission
                                    *_configMaker);
 }
 
+- (void)createOfferForRealmAtURL:(NSURL *)url
+                     accessLevel:(RLMSyncAccessLevel)accessLevel
+                      expiration:(NSDate *)expirationDate
+                        callback:(RLMPermissionOfferStatusBlock)callback {
+    if (!_user || _user->state() == SyncUser::State::Error) {
+        callback(nil, make_permission_error_change(@"A permission offer cannot be created using an invalid user."));
+        return;
+    }
+    auto cb = [callback](util::Optional<std::string> token, std::exception_ptr ptr) {
+        if (ptr) {
+            NSError *error = translateSyncExceptionPtrToError(std::move(ptr), RLMPermissionActionTypeOffer);
+            REALM_ASSERT(error);
+            callback(nil, error);
+        } else {
+            REALM_ASSERT_DEBUG(token != nullptr);
+            callback([NSString stringWithUTF8String:token->c_str()], nil);
+        }
+    };
+    auto offer = PermissionOffer{
+        [tildeSubstitutedPathForRealmURL(url, self.identity) UTF8String],
+        accessLevelForObjcAccessLevel(accessLevel),
+        RLMTimestampForNSDate(expirationDate),
+    };
+    Permissions::make_offer(_user, std::move(offer), std::move(cb), *_configMaker);
+}
+
+- (void)acceptOfferForToken:(NSString *)token
+                   callback:(RLMPermissionOfferResponseStatusBlock)callback {
+    if (!_user || _user->state() == SyncUser::State::Error) {
+        callback(nil, make_permission_error_change(@"A permission offer cannot be accepted by an invalid user."));
+        return;
+    }
+    auto cb = [callback](util::Optional<std::string> raw_url, std::exception_ptr ptr) {
+        if (ptr) {
+            NSError *error = translateSyncExceptionPtrToError(std::move(ptr), RLMPermissionActionTypeAcceptOffer);
+            REALM_ASSERT(error);
+            callback(nil, error);
+        } else {
+            REALM_ASSERT_DEBUG(bool(raw_url));
+            NSString *urlAsString = [NSString stringWithUTF8String:raw_url->c_str()];
```

Comment:
`@(raw_url->c_str())`, which is probably concise enough you can do away with the extra local variable.